Miracast/P2P/MiracastP2P.cpp

```diff
@@ -0,0 +1,546 @@
+    }
+
+    ret = wpa_ctrl_request(wpa_p2p_ctrl_iface, cmd, strlen(cmd), ret_buf, &buf_len, NULL);
+
+    if (ret == -2)
+    {
+        MIRACASTLOG_ERROR("WIFI_HAL: cmd=%s timed out ", cmd);
+    }
+    else if (ret < 0)
+    {
+        MIRACASTLOG_ERROR("WIFI_HAL: cmd=%s failed ", cmd);
+    }
+    MIRACASTLOG_TRACE("Exiting...");
+    return ret;
+}
+
+// Initializes WiFi - P2P
+// Connects to the wpa_supplicant via control interface
+// Gets attached to wpa_supplicant to receiver events
+// Starts the p2p_monitor thread
+MiracastError MiracastP2P::p2pInit()
+{
+    int retry = 0;
+    m_stop_p2p_monitor = false;
+    pthread_attr_t thread_attr;
+    int ret = 0;
+    MIRACASTLOG_TRACE("Entering..");
+    MIRACASTLOG_INFO("WIFI_HAL: Initializing P2P WiFi HAL.");
+    const char* wpa_supp_ctrl_path_ref_file = "/opt/mcast_wpa_ctrl_path.txt";
+    std::string wpa_supp_ctrl_path_name;
+    std::ifstream wpa_supp_ctrl_path_ref_fs(wpa_supp_ctrl_path_ref_file);
+
+    if (wpa_supp_ctrl_path_ref_fs.is_open())
+    {
+        std::getline(wpa_supp_ctrl_path_ref_fs, wpa_supp_ctrl_path_name);
+        MIRACASTLOG_INFO("WPA Supplicant ctrl path reading from file [%s], wpa_supp_ctrl_path as [ %s] ", wpa_supp_ctrl_path_ref_file, wpa_supp_ctrl_path_name.c_str());
+        wpa_supp_ctrl_path_ref_fs.close();
+    }
+    else{
+        wpa_supp_ctrl_path_name = WPA_SUP_DFLT_CTRL_PATH;
+    }
+
+    if (0 != access(wpa_supp_ctrl_path_name.c_str(), F_OK)){
+        MIRACASTLOG_ERROR("Unable to find P2P ctrl iface path[%s]", wpa_supp_ctrl_path_name.c_str());
+        return MIRACAST_INVALID_P2P_CTRL_IFACE;
+    }
+
+    /* Starting wpa_supplicant may take some time, try 10 times before giving up */
+    retry = 0;
+    while (retry++ < 10)
+    {
+        m_wpa_p2p_cmd_ctrl_iface = wpa_ctrl_open(wpa_supp_ctrl_path_name.c_str());
+        if (m_wpa_p2p_cmd_ctrl_iface != NULL)
+            break;
+        MIRACASTLOG_ERROR("WIFI_HAL: p2p ctrl_open returned NULL ");
+        sleep(1);
+    }
+    if (m_wpa_p2p_cmd_ctrl_iface == NULL)
+    {
+        MIRACASTLOG_ERROR("WIFI_HAL: wpa_ctrl_open for p2p failed for control interface ");
+        MIRACASTLOG_TRACE("Exiting...");
+        return MIRACAST_P2P_INIT_FAILED;
+    }
+    MIRACASTLOG_INFO("WIFI_HAL: m_wpa_p2p_cmd_ctrl_iface created successfully.");
+
+    m_wpa_p2p_ctrl_monitor = wpa_ctrl_open(wpa_supp_ctrl_path_name.c_str());
+    if (m_wpa_p2p_ctrl_monitor == NULL)
+    {
+        MIRACASTLOG_ERROR("WIFI_HAL: wpa_ctrl_open for p2p failed for monitor interface ");
+        wpa_ctrl_close(m_wpa_p2p_cmd_ctrl_iface);
+        m_wpa_p2p_cmd_ctrl_iface = nullptr;
+        MIRACASTLOG_TRACE("Exiting...");
+        return MIRACAST_P2P_INIT_FAILED;
+    }
+    MIRACASTLOG_INFO("WIFI_HAL: m_wpa_p2p_ctrl_monitor created successfully.");
+    if (wpa_ctrl_attach(m_wpa_p2p_ctrl_monitor) != 0)
+    {
+        MIRACASTLOG_ERROR("WIFI_HAL: p2p wpa_ctrl_attach failed ");
+        wpa_ctrl_close(m_wpa_p2p_cmd_ctrl_iface);
```

Comment:
we can write a utlity function to avoid repeataion

Reply:
Addressed the review comments in https://github.com/rdkcentral/rdkservices/pull/4618/commits/9e283d757407fc8e7010eaed21427ea2af2865d6